### ISSUE TITLE
MCS: Prevent budget amplification via IPC calls to passive servers

### ISSUE THREAD

**Issue by Indanz** [seL4/seL4#1365]
Currently it is theoretically possible to indefinitely starve other equal priority tasks.

Opening this issue before I forget about it. From my mattermost discussion about #902 with Gernot:

Another thing that affects fairness in cpu time between equal priority tasks are calls to higher priority passive servers: If the budget runs out during such call, that task will effectively replenish its budget and keep running, because at the time it ran out, it was highest priority. This is a much smaller practical problem than the signal issue, but theoretically, if timing is very unfortunate, one task could starve another equal priority task indefinitely this way. (One solution would be to remember if this happened, and if it did, append the task at the end of the queue when returning the SC. But this seems tricky to implement.)

(Though maybe it's as simple as remembering the priority inside the SC when the last replenishment was done, and checking that when returning the SC to the caller.)

Unlike #902, this problem is more theoretical and not likely to lead to problems in real systems under normal workloads. However, it is easy to trigger by malicious processes.

Simplest solution to this that I can think of is to remember the task's priority when the budget gets replenished. Then when unblocking the caller, its priority is compared to the saved one in the SC:

- If it's the same or higher than the saved priority, proceed as usual.
- Otherwise append the task to the end of the run queue instead of the head.

Timeout faults do not help because callers can be round-robin tasks.

**Comment by gernotheiser** [comment]
I don't claim I have thought thought this in detail, but on the face of it I agree that this looks wrong

**Comment by Indanz** [comment]
A further improvement would be to subtract the amount of budget overrun from the caller's budget.

This doesn't fit in how round-robin tasks are currently implemented though, as it either requires a negative budget or a way to postpone a round-robin task, but worth looking into.

If implemented this way, then tasks can only cause temporary budget overruns and overall can't amplify their budget at all any more.

However, the added advantage compared to putting it at the back of the queue is limited:
- If there are no other tasks waiting, there is no reason not to allow this budget overrun, because the task would have gotten it anyway, hence in hindsight it's not an overrun. Unlike sporadic tasks, round-robin task don't have a set bandwidth, but what it should be depends on other equal priority tasks' state.
- If there are other tasks, putting it at the end of the queue gives everyone else a chance to run. Only if the server needs to replenish the budget more than once during an IPC call does it start to make a practical difference. The added complexity of handling this corner case may not be worth it.
